lib/bridge/MemberListSyncer.js

```diff
@@ -247,10 +271,63 @@ function joinMatrixUsersToChannels(rooms, server, injectJoinFn) {
     return d.promise;
 }
 
-function leaveIrcUsersFromRooms(rooms, server) {
+MemberListSyncer.prototype.leaveIrcUsersFromRooms = function(rooms, server) {
+    log.info("leaveIrcUsersFromRooms");
+
+    // Store the matrix room info in memory for later retrieval when NAMES is received
+    // and updateIrcMemberList is called. At that point, we have enough information to
+    // leave users from the channel that the NAMES is for.
+    rooms.forEach((roomInfo) => {
+        this._memberLists.matrix[roomInfo.id] = roomInfo;
+    });
+
     return Promise.resolve();
 }
 
+// Update the MemberListSyncer with the IRC NAMES_RPL that has been received for channel.
+// This will leave any matrix users that do not have their associated IRC nick in the list
+// of names for this channel.
+MemberListSyncer.prototype.updateIrcMemberList = Promise.coroutine(function*(channel, names) {
+    log.info(`updateIrcMemberList: Updating IRC member list for ${channel}`);
```

Comment:
Careful. The bridge is [configurable](https://github.com/matrix-org/matrix-appservice-irc/blob/master/config.sample.yaml#L170) as to whether this should be done, and you aren't checking for it.

It might be an idea to also log the *number* of names so we have some knowledge whether we're about to kill Synapse or not :D